hdr/sqlite_modern_cpp.h

```diff
@@ -288,31 +209,193 @@ namespace sqlite {
+		template <typename ...Args>
+		struct is_sqlite_value< std::variant<Args...> > : public std::integral_constant<
+			bool,
+			true
+		> { };
+#endif
+#ifdef MODERN_SQLITE_STD_OPTIONAL_SUPPORT
+		template <typename T>
+		struct is_sqlite_value< optional<T> > : public is_sqlite_value<T> {};
+#endif
+
+#ifdef _MODERN_SQLITE_BOOST_OPTIONAL_SUPPORT
+		template <typename T>
+		struct is_sqlite_value< boost::optional<T> > : public is_sqlite_value<T> {};
+#endif
+	}
+
+	class row_iterator {
+	public:
+		class value_type {
+		public:
+			value_type(database_binder *_binder): _binder(_binder) {};
+			template<class T>
+			typename std::enable_if<detail::is_sqlite_value<T>::value, value_type &>::type operator >>(T &result) {
+				get_col_from_db(*_binder, next_index++, result);
+				return *this;
+			}
+			template<class ...Types>
+			value_type &operator >>(std::tuple<Types...>& values);
+			template<class ...Types>
+			value_type &operator >>(std::tuple<Types...>&& values) {
+				return *this >> values;
+			}
+			template<class ...Types>
+			operator std::tuple<Types...>() {
+				std::tuple<Types...> value;
+				*this >> value;
+				return value;
+			}
+			explicit operator bool() {
+				return sqlite3_column_count(_binder->_stmt.get()) >= next_index;
+			}
+		private:
+			database_binder *_binder;
+			int next_index = 0;
+		};
+		using difference_type = std::ptrdiff_t;
+		using pointer = value_type*;
+		using reference = value_type&;
+		using iterator_category = std::input_iterator_tag;
+
+		row_iterator() = default;
+		explicit row_iterator(database_binder &binder): _binder(&binder) {
+			_binder->_next_index();
+			_binder->_inx = 0;
+			_binder->used(true);
+			++*this;
+		}
+
+		reference operator*() const { return value;}
+		pointer operator->() const { return std::addressof(**this); }
+		row_iterator &operator++() {
+			switch(int result = sqlite3_step(_binder->_stmt.get())) {
+				case SQLITE_ROW:
+					value = {_binder};
+					break;
+				case SQLITE_DONE:
+					_binder = nullptr;
+					break;
+				default:
+					exceptions::throw_sqlite_error(result, _binder->sql());
+			}
+			return *this;
 		}
 
-		template<typename... Types>
-		void operator>>(std::tuple<Types...>&& values) {
-			this->_extract_single_value([&values, this] {
-				tuple_iterate<std::tuple<Types...>>::iterate(values, *this);
-			});
+		// Not well-defined
+		row_iterator operator++(int);
+		friend inline bool operator ==(const row_iterator &a, const row_iterator &b) {
+			return a._binder == b._binder;
+		}
+		friend inline bool operator !=(const row_iterator &a, const row_iterator &b) {
+			return !(a==b);
 		}
 
-		template <typename Function>
-		typename std::enable_if<!is_sqlite_value<Function>::value, void>::type operator>>(
-			Function&& func) {
-			typedef utility::function_traits<Function> traits;
+	private:
+		database_binder *_binder = nullptr;
+		mutable value_type value{_binder}; // mutable, because `changing` the value is just reading it
```

Comment:
@zauguin We have no const member function assigning to `value`, why do we need the `mutable` ?

Reply:
If this is not mutable, operator* has to return a const reference  to `value_type` instead.
But reading from the `value_type` requires increasing the index for the next column, `next_index`.

I see four ways to solve this:

1. Make `operator *` and `operator ->` not const.
    This breaks requirements for iterators and might break code. IMO it would also be weird, because these functions are not changing anything.
2. Make `reference` the same as `value_type` and return a new `value_type` at every call to the operators.
    If this is ever used with a not range for loop, everytime `*iter` is used the read restarts at the first column.
    So this could lead to very hard to find bugs.
3. Make the `value_type` member functions `const` and `value_type::next_index` mutable. Then a const reference can be returned.
    This would work, but then we work `const` qualified instances of value_type, which don't act like `const` objects at all because they change if any function on them is called.
4. The current approach: A mutable `value`.
    This is certainly not perfect, but it moves the problems of 3. to the iterator type which is less visible to the user.

Reply:
@zauguin Thanks for the clarification, Since `value_type` does not own the underlying data, I too think solution No 4 is the right choice here.

Right now `value_type` will only read data if dereferenced (`*value`) or when it is implicitly converted to `std::tuple`, basically a non-owing type.

What are your thoughts if we make it own the entire row?
`value_type` can own an `std::tuple` member, or drive from `std::tuple<Types...>`
We will initialize & update the underlying tuple in `row_iterator::operator++`.

Reply:
@zauguin If we drive from an `std::tuple`, can we take advantage of C++17 structured bindings?

Reply:
@aminroosta This sounds like the interface of #103.
The fundamental problem is determining the type and number of columns at compile time. If the value_type should own the values, the value_type has to know about <Types...>, so the types have to be passed as explicit template arguments.

Reply:
@aminroosta How do you prefere to move forward here? Do you prefer the interface opions discussed in #103 or should we continue with this one?